api/proto/auth/config.go

```diff
@@ -0,0 +1,97 @@
+package auth
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+)
+
+// Config contains configuration of the client
+type Config struct {
+	// Addrs is a list of addresses to dial
+	Addrs []string
+	// Dialer is a custom dialer, if provided
+	// is used instead of the list of addresses
+	Dialer ContextDialer
+	// KeepAlivePeriod defines period between keep alives
+	KeepAlivePeriod time.Duration
+	// KeepAliveCount specifies amount of missed keep alives
+	// to wait for until declaring connection as broken
+	KeepAliveCount int
+	// TLS is a TLS config
+	TLS *tls.Config
+}
+
+// CheckAndSetDefaults checks and sets default config values
+func (c *Config) CheckAndSetDefaults() error {
+	if len(c.Addrs) == 0 && c.Dialer == nil {
+		return trace.BadParameter("set parameter Addrs or DialContext")
+	}
+	if c.TLS == nil {
+		return trace.BadParameter("missing parameter TLS")
+	}
+	if c.KeepAlivePeriod == 0 {
+		c.KeepAlivePeriod = ServerKeepAliveTTL
+	}
+	if c.KeepAliveCount == 0 {
+		c.KeepAliveCount = KeepAliveCountMax
+	}
+	if c.Dialer == nil {
+		c.Dialer = NewAddrDialer(c.Addrs, c.KeepAlivePeriod)
+	}
+	if c.TLS.ServerName == "" {
+		c.TLS.ServerName = teleport.APIDomain
+	}
+	// this logic is necessary to force client to always send certificate
+	// regardless of the server setting, otherwise client may pick
+	// not to send the client certificate by looking at certificate request
+	if len(c.TLS.Certificates) != 0 {
+		cert := c.TLS.Certificates[0]
+		c.TLS.Certificates = nil
+		c.TLS.GetClientCertificate = func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return &cert, nil
+		}
+	}
+
+	return nil
+}
+
+// ContextDialer represents network dialer interface that uses context
+type ContextDialer interface {
```

Comment:
Nit: Not sure this and everything below belongs to the config.go module, I'd consider moving to contextdialer.go maybe?